model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java

```diff
@@ -40,12 +40,24 @@
  */
 public class MapUserProviderFactory extends AbstractMapProviderFactory<MapUserProvider, MapUserEntity, UserModel> implements UserProviderFactory<MapUserProvider>, InvalidationHandler {
 
+    public static final String USERNAME_CASE_SENSITIVE = "username-case-sensitive";
+
     public MapUserProviderFactory() {
         super(UserModel.class, MapUserProvider.class);
     }
 
     @Override
     public MapUserProvider createNew(KeycloakSession session) {
+        RealmModel realm = session.getContext().getRealm();
+        boolean usernameCaseSensitive = false; //default to false
+
+        // if a realm is available take it from there
+        if (realm != null) {
+            usernameCaseSensitive = realm.getAttribute(USERNAME_CASE_SENSITIVE, Boolean.FALSE);
+        }
+
+        session.setAttribute(USERNAME_CASE_SENSITIVE, usernameCaseSensitive);
```

Comment:
> Why do we configure case-sensitiveness per realm? 

The requirement comes from the discussion. See https://github.com/keycloak/keycloak/discussions/10198#discussioncomment-2179669 for the reference.

> What if someone disables case-sensitiveness when storage contains conflicting usernames? Should we have this switch somehow more static/fixed as part of the storage config maybe?

I asked similar question here: https://github.com/keycloak/keycloak/pull/10332#issuecomment-1169172848 

> Why are we storing this value in `session`? If case-sensitiveness is configured per realm, what if we do user search per more realms within one `session`? For example, search admin user in realm1 and then realm2? Can we just do `realm.getAttribute` in `MapUserProvider`?

Good point. We've discussed it with @hmlnarik in one of the sync calls and agreed that the session will be used. I don't recall exactly the reason right now, but it's true it won't work as it now with more realms within one session. I was able to reproduce it in model test. Right now I'm thinking whether it can happen also in 'real' scenario, when you use rest endpoints. Anyway I'm happy to update it. @hmlnarik could you please comment on this? Thank you

Reply:
Thanks for the explanation @vramik, sorry I missed the discussion points. Let's wait for Hynek for feedback.

Reply:
Unfortunately there is not always appropriate realm in the context. I was able to set the correct realm at some places, but then I discovered a scenario where it wouldn't be easy. E.g. during importing a realm there is 

1. realm admin request authenticated for "master" realm where identity token is being verified for "admin" user to check whether the user has permission to create new realms
3. it leads to a creation of new instance of `UserProvider` (`MapUserProvider`) within the session, where case-sensitivity attribute for "master" realm would be added into the session attributes
4. Then when request lands into `RealmsAdminResource.importRealm(RealmRepresentation)` trying to create new realm, there is already provider created in the session, so no new provider is being created where it could be the attribute for the new realm inserted into session
5. when new users are tried to be created it's checked whether the user exist already and this call fail due to the above

One possible solution could be not to create new provider only within the session but also for each realm. 

As discussed with @mhajas ... for now I'll move the logic into `MapUserProvider` only and check directly the realm attribute only when it's needed for the particular search.

Reply:
Thanks for the investigation @vramik. Personally, I like the solution this way. It would make sense to have it in the session if the realm is always the same within each request, but since it is not the case, as you found out, I lean towards the solution you proposed in the latest changes.

If we want to optimize the obtaining of the lettercase flag from the realm, maybe we should take a step back and think about deploying some generic solution for all realm attributes. WDYT? 

Some ideas already here: https://github.com/keycloak/keycloak/discussions/12328

Reply:
> If we want to optimize the obtaining of the lettercase flag from the realm, maybe we should take a step back and think about deploying some generic solution for all realm attributes. WDYT?
> 
> Some ideas already here: #12328

Thanks, agreed.